src/Microsoft.DotNet.Darc/src/Darc/Operations/CloneOperation.cs

```diff
@@ -28,25 +31,134 @@ public override async Task<int> ExecuteAsync()
+                                {
+                                    Logger.LogWarning($"Skipping dependency from {repo.RepoUri}@{repo.Commit} to {d.RepoUri}: Missing commit.");
+                                }
+                                else
+                                {
+                                    StrippedDependency stripped = new StrippedDependency(d);
+                                    if (!seenDependencies.Contains(stripped))
+                                    {
+                                        Logger.LogDebug($"Adding new dependency {stripped.RepoUri}@{stripped.Commit}");
+                                        seenDependencies.Add(stripped);
+                                        accumulatedDependencies.Add(stripped);
+                                    }
+                                }
+                            });
+                            // delete the .gitdir redirect to orphan the repo.
+                            // we want to do this because otherwise all of these folder will show as dirty in Git,
+                            // and any operations on them will affect the master copy and all the others, which
+                            // could be confusing.
+                            string repoGitRedirectPath = Path.Combine(repoPath, ".git");
+                            if (File.Exists(repoGitRedirectPath))
+                            {
+                                File.Delete(repoGitRedirectPath);
+                            }
+                        }
+                        catch (DirectoryNotFoundException)
+                        {
+                            Logger.LogWarning($"Repo {repoPath} appears to have no '/eng' directory at commit {repo.Commit}.  Dependency chain is broken here.");
+                        }
+                        catch (FileNotFoundException)
+                        {
+                            Logger.LogWarning($"Repo {repoPath} appears to have no '/eng/Version.Details.xml' file at commit {repo.Commit}.  Dependency chain is broken here.");
+                        }
+
+                        Logger.LogDebug($"Now have {dependenciesToClone.Count} dependencies to consider");
+                    }   // end inner while(dependenciesToClone.Any())
+
+
+                    if (_options.CloneDepth == 0 && accumulatedDependencies.Any())
+                    {
+                        Logger.LogInformation($"Reached clone depth limit, aborting with {accumulatedDependencies.Count} dependencies remaining");
+                        foreach (StrippedDependency d in accumulatedDependencies)
+                        {
+                            Logger.LogDebug($"Abandoning dependency {d.RepoUri}@{d.Commit}");
+                        }
+                        break;
+                    }
+                    else
+                    {
+                        _options.CloneDepth--;
```

Comment:
Do expect that this will converge in a typical graph? I don't think it will at times.  For instance, if there is a cycle in the graph (DotNet-Trusted and core-setup have a cycle today)

Reply:
Think I have this handled right now, I'm testing it out.

---
In reply to: [275061860](https://github.com/dotnet/arcade-services/pull/326#discussion_r275061860) [](ancestors = 275061860)